source/common/http/conn_manager_impl.cc

```diff
@@ -37,6 +37,31 @@
 namespace Envoy {
 namespace Http {
 
+namespace {
+
+template <class T> using FilterList = std::list<std::unique_ptr<T>>;
+
+// Shared helper for recording the latest filter used.
+template <class T>
+void recordLatestDataFilter(const typename FilterList<T>::iterator current_filter,
+                            T** latest_filter, const FilterList<T>& filters) {
+  // If this is the first time we're calling onData, just record the current filter.
+  if (*latest_filter == nullptr) {
+    *latest_filter = current_filter->get();
+    return;
+  }
+
+  // We want to keep this pointing at the latest filter in the filter list that has received the
+  // onData callback. To do so, we compare the current latest with the *previous* filter. If they
+  // match, then we must be processing a new filter for the first time. We omit this check if we're
+  // the first filter, since the above check handles that case.
+  if (current_filter != filters.begin() && *latest_filter == std::prev(current_filter)->get()) {
```

Comment:
Why not just universally set latest to current? Seems simple to read/reason about?

Reply:
I wanted to cover the case where there are multiple onData callbacks: If we just set latest to current, then the first onData filter iteration would correctly iterate over the the filters and set `latest`, but on subsequent onData iterations we'd start from the beginning again, potentially allowing filter N to modify the buffer even though filter M > N was the filter that inserted data into the buffer. 

Hopefully this makes sense - open for suggestions if this seems unnecessary or if there's a better way.

Reply:
I see OK, makes sense. Can you clarify that a bit in the above comment (just add your additional explanation in the above comment)?